### ISSUE TITLE
Migration to python 3.12

### ISSUE THREAD

**Issue by drew2a** [Tribler/tribler#7282]
Python 3.11 offers some performance improvements that are good to have: https://docs.python.org/3/whatsnew/3.11.html#faster-runtime.

Python 3.11 also is a must-have for all Apple Silicone owners if they want to build Tribler easily: https://github.com/Tribler/tribler/issues/7252 https://github.com/Tribler/tribler/pull/7281

Currently, the following libraries from Tribler's dependencies are not ready for this migration:
- [x] ipv8: https://github.com/Tribler/py-ipv8/issues/1113
- [x] ponyORM: https://github.com/ponyorm/pony/issues/668
- [x] libtorrent: https://github.com/arvidn/libtorrent/discussions/7244 https://github.com/arvidn/libtorrent/issues/7536

**Comment by ichorid** [comment]
The real problem is Libtorrent: building it with Python 3.11 requires Boost 1.81, which was published only recently. However, the PIP version of Libtorrent is pretty old and does not support 3.11.

Related discussions:
[Instructions for building Libtorrent with Python 3.11 bindings](https://github.com/arvidn/libtorrent/discussions/7187)

["update pypi libtorrent"](https://github.com/arvidn/libtorrent/discussions/7244)

["Who is responsible for which Python PIP package?"](https://github.com/arvidn/libtorrent/discussions/7178)

**Comment by drew2a** [comment]
@ichorid thank you for the information! I've updated the issue's description.

**Comment by Solomon1732** [comment]
I'll add that Python 3.10 added match statements: [What's new in python 3.10](https://docs.python.org/3/whatsnew/3.10.html), [specification](https://peps.python.org/pep-0634/), [motivations and rationale](https://peps.python.org/pep-0635/), [tutorial](https://peps.python.org/pep-0636/). It can negate the need for many `if-elif` chains when writing code.

**Comment by Solomon1732** [comment]
There looks to be a difference between [this](https://github.com/arvidn/libtorrent/blob/2b6360320b1b1bf3a28b72c9866071b2b155ac21/pyproject.toml#L5) somewhat older libtorrent version and [this](https://github.com/arvidn/libtorrent/blob/RC_2_0/pyproject.toml) 2.0 branch. Namely, python 3.11 isn't skipped anymore. That said, I don't know if said branch is a stable release or a release candidate. Its release page isn't clear about it:
https://github.com/arvidn/libtorrent/releases

**Comment by Solomon1732** [comment]
Done. Hopefully I'll be answered Soon:tm:: https://github.com/arvidn/libtorrent/issues/7265#issuecomment-1627030797

**Comment by Solomon1732** [comment]
Yes, it's stable: https://github.com/arvidn/libtorrent/issues/7265#issuecomment-1634090403

**Comment by drew2a** [comment]
From what I see, Python 3.11 support isn't available in either [v2.0.9](https://github.com/arvidn/libtorrent/blob/v2.0.9/pyproject.toml#L5) or [v1.2.19](https://github.com/arvidn/libtorrent/blob/v1.2.19/pyproject.toml#L5).

Support for Python 3.11 is only present in [RC_2_0](https://github.com/arvidn/libtorrent/blob/RC_2_0/pyproject.toml#L5).

You can find the specific commit [here](https://github.com/arvidn/libtorrent/commit/a43e7f718a667fb1a4a2aaeed9061fcff154e6ba#diff-62587956f943bb2503db7bc6dd27d0d888074a1c0ecaab3f570ad611aff0f7bbR48).

**Comment by Solomon1732** [comment]
I missed it. Thanks. It's a shame. Hopefully it'll be be supported soon.

**Comment by kozlovsky** [comment]
PonyORM 0.7.17 with Python 3.11 support is out! So, what's left is libtorrent

**Comment by Solomon1732** [comment]
libtorrent version 2.0.10 came out in February. It looks like it supports Python 3.11 now? I am wondering tentatively. Here's the release: https://github.com/arvidn/libtorrent/releases/tag/v2.0.10

One of the notes says this:
> Bump boost version to enable Python 3.11 package builds by @cas-- in https://github.com/arvidn/libtorrent/pull/7445

Also, if tribler migrates to Python 3.11, does it mean issues like #7869 can be closed, if only to tidy things up?

**Comment by drew2a** [comment]
@Solomon1732, thank you for noticing this and keeping us informed.

Currently, we use libtorrent 1.2.x in Tribler. Switching to 2.x is still under debate. For more information, please check:
- #5556

**Comment by RealJosephKnapp** [comment]
Perhaps it is time to upgrade to Libtorrent 2.0.10, and to upgrade its dependencies to Boost 1.86, zlib 1.3.1 and OpenSSL 3.3.2.

**Comment by qstokkink** [comment]
🎉 libtorrent `2.0.11` has been released with modern Python support.
